gRPC Storage Service

VineethReddy02 opened the pull request:
**What this PR does**:
This PR contains gRPC service client i.e
1. Index Client
2. Storage Client
3. Table Client

This is still in work in progress PR this needs testing with gRPC storage server implementation

**Which issue(s) this PR fixes**:
Fixes #1681

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+
+package grpc;
+
+service grpc {
```
grpc-store

VineethReddy02 replied:
.protoc file doesn't accept **grpc-store** So named it as grpc_store

gouthamve commented on docs/configuration/config-file-reference.md:
```diff
@@ -1551,6 +1551,11 @@ delete_store:
   # CLI flag: -deletes.requests-table-name
   [requests_table_name: <string> | default = "delete_requests"]
+
+grpc:
```
grpc-store

VineethReddy02 replied:
Renamed it.

gouthamve commented on pkg/chunk/grpc/index_client.go:
```diff
@@ -0,0 +1,108 @@
+	"github.com/cortexproject/cortex/pkg/chunk/util"
+)
+
+type writeBatch struct {
```
Why do we have this. Why can't we use `WriteBatch` directly?

VineethReddy02 replied:
I see Cassandra is maintaining the writeBatch implicitly within itself. So, I also added writeBatch within grpc client. I see dynamoDBstorageclient maintaining one within itself.

gouthamve replied:
Yes, but we have `grpc.WriteBatch` already. Why are we using another one? Specially when we're just copying this in `BatchWrite()`

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+service grpc {
+    //index-client
+    rpc BatchWrite(WriteBatch) returns (Nothing);
+    rpc QueryPages(IndexQuery) returns (ReadBatch);
```
We shouldn't put the Cassandra limitation of a single query at once into the grpc layer. Should be `rpc QueryPages([]IndexQuery) returns (ReadBatch);`, which should return a streaming response, with each read on the stream returning a `ReadBatch`.

VineethReddy02 replied:
@gouthamve Made required changes to support multiple rows per stream.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+    repeated IndexEntry indexEntry = 1;
+}
+
+message ReadBatch {
```
Why is the ReadBatch only returning a single row? We should accept multiple rows in the ReadBatch.

VineethReddy02 replied:
Changes this to stream. So now ReadBatch sends multiple rows in the single RPC call

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+    rpc QueryPages(IndexQuery) returns (ReadBatch);
+
+    //storage-client
+    rpc PutChunks(Chunk) returns (Nothing);
```
PutChunks() should accept multiple chunks not a single chunk.

VineethReddy02 replied:
Made the necessary changes to do PutChunks multiple chunks in a single RPC call.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+    int64 provisionedWrite       = 4;
+    map<string, string> tags     = 5;
+    AutoScalingConfig writeScale = 6;
+    AutoScalingConfig readScale  = 7;
```
All these are dynamo specific, we can remove these.

VineethReddy02 replied:
Yeah. Removed them.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,121 @@
+    int64 outCoolDown   = 5;
+    int64 inCoolDown    = 6;
+    double targetValue   = 7;
+}
```
Remove this

VineethReddy02 replied:
Removed this.

VineethReddy02 commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,122 @@
+
+service grpc {
+    //index-client
+    rpc NewWriteBatch(Nothing) returns (WriteBatch);
```
Removed it.

VineethReddy02 commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,127 @@
+}
+
+message ReadBatch {
+   repeated Row rows = 1;
```
@gouthamve added as you mentioned.

gouthamve commented on docs/configuration/config-file-reference.md:
```diff
@@ -1574,6 +1574,11 @@ filesystem:
 
+grpc-store:
+  # Hostname or IP of gRPC instance.
+  # CLI flag: -grpc.server-addresses
```
`-grpc-store.server-address`?

VineethReddy02 replied:
Changed the flag name as suggested.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,127 @@
+}
+
+message ChunkResponse {
+    uint32 checksum = 1;
```
Why are we returning the checksum here? Further, I checked the Cassandra grpc implementation code and you're just returning the same checksum as the caller?

VineethReddy02 replied:
@gouthamve I am returning the same checksum. Shall I drop it from the response?

VineethReddy02 replied:
Dropped the checksum here.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,127 @@
+
+    //storage-client
+    rpc PutChunks(ChunksData) returns (Nothing);
+    rpc GetChunks(Chunks) returns (ChunksResponse);
```
This doesn't work yet. This requires that the upstream service has knowledge about the schema (to figure out which table to read from), but that shouldn't be the case. We should send the table info as well.

VineethReddy02 replied:
Yeah fixed it. gRPC server will not be aware of `schemaConfig`. `Tablename `will be sent in `GetChunks `request field.

gouthamve commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,127 @@
+
+    //storage-client
+    rpc PutChunks(ChunksData) returns (Nothing);
+    rpc GetChunks(Chunks) returns (ChunksResponse);
```
Needs to be a streaming response as we can return GBs of data

VineethReddy02 commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,127 @@
+    rpc UpdateTable(UpdateTableRequest) returns (Nothing);
+
+    //used by storage-client & index-client
+    rpc Stop(Nothing) returns (Nothing);
```
I personally feel let it be in proto. I also added a gRPC server call at gRPC client Stop() invocation.
In server-level am closing the DB connection.

Having this in spec indicates we have added all the calls existing in cortex spec.

VineethReddy02 commented on pkg/chunk/grpc/index_client.go:
```diff
@@ -0,0 +1,103 @@
+	b := batch.(*WriteBatch)
+	writeBatch := &WriteBatch{}
+	writeBatch.IndexEntry = []*IndexEntry{}
+	for _, entry := range b.IndexEntry {
```
I missed this removing the copying part of code.

gouthamve commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,121 @@
+				var c chunk.Chunk
+				if chunkResponse != nil {
+					decodeContext := chunk.NewDecodeContext()
+					// if error ignore as partial results are still useful
```
We should still return the error.

Can we have
```
err := c.Decode(decodeContext, chunkResponse.Buf)
if err != nil {
  lastErr = err
}
```

And return `lastErr`

We can also reuse `decodeContext` by creating at the top and not have to create it again and again.

pracucci replied:
My only comment is that `firstErr` could be more useful than `lastErr`. It's generally preferable seeing the first error not the last error. The last error could be a consequence of first error, but no viceversa.

gouthamve commented on pkg/chunk/grpc/index_client.go:
```diff
@@ -59,8 +59,10 @@ func (s *StorageClient) query(ctx context.Context, query chunk.IndexQuery, callb
 		}
-		if !callback(readBatch) {
-			return nil
+		if readBatch != nil {
```
We should check the error here tbh.

pracucci replied:
What do you mean @gouthamve? The error is checked above. However, I'm not sure why we need this `if` at all.

VineethReddy02 replied:
@pracucci Though we check the error. If the response is nil the callback func fails. I am just trying to make sure if grpc server doesn't handle all cases. cortex shouldn't crash in any scenario.

pracucci replied:
If the gRPC server is buggy and returns `nil` batch and `nil` error (which is a case we never expect to happen) you will end up silently ignoring it and you would potentially end up with wrong query results. If you would have the same exact bug in Cortex today, Cortex would panic. I don't see why we should do it differently with the gRPC store.

VineethReddy02 replied:
Okay. In that case, I can remove the additional ```nil``` value validation.

pracucci commented:
_I've pushed a commit to rebase `master`._

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+syntax = "proto3";
```
I would keep consistency with filenames and rename `gRPC.proto` to the lower case `grpc.proto`. Then remember to change it in `Makefile` too.

pracucci commented on pkg/chunk/storage/factory.go:
```diff
@@ -57,20 +59,18 @@ type StoreLimits interface {
+	BoltDBConfig            local.BoltDBConfig       `yaml:"boltdb"`
+	FSConfig                local.FSConfig           `yaml:"filesystem"`
+	IndexCacheValidity      time.Duration            `yaml:"index_cache_validity"`
+	GrpcConfig              grpc.Config              `yaml:"grpc-store"`
```
The convention is `_` in YAML, `-`  in CLI flags:

```suggestion
	GrpcConfig              grpc.Config              `yaml:"grpc_store"`
```

pracucci commented on pkg/chunk/grpc/grpc_client.go:
```diff
@@ -0,0 +1,25 @@
+
+// RegisterFlags adds the flags required to config this to the given FlagSet
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.Address, "grpc-store.server-address", "", "Hostname or IP of gRPC instance.")
```
```suggestion
	f.StringVar(&cfg.Address, "grpc-store.server-address", "", "Hostname or IP of the gRPC store instance.")
```

pracucci commented on pkg/chunk/grpc/grpc_client.go:
```diff
@@ -0,0 +1,25 @@
+
+// Config for a StorageClient
+type Config struct {
+	Address string `yaml:"address,omitempty"`
```
We should keep naming consistency between YAML and CLI flags. Either we rename this to `server_address` or the CLI flag to `grpc-store.address`.

pracucci commented on pkg/chunk/grpc/grpc_client.go:
```diff
@@ -0,0 +1,25 @@
+func connectToGrpcServer(serverAddress string) (GrpcStoreClient, *grpc.ClientConn, error) {
+	cc, err := grpc.Dial(serverAddress, grpc.WithInsecure())
+	if err != nil {
+		return nil, nil, err
```
```suggestion
		return nil, nil, errors.Wrapf(err, "failed to dial grpc-store %s", serverAddress)
```

pracucci commented on pkg/chunk/grpc/grpc_client.go:
```diff
@@ -0,0 +1,25 @@
+	f.StringVar(&cfg.Address, "grpc-store.server-address", "", "Hostname or IP of gRPC instance.")
+}
+
+func connectToGrpcServer(serverAddress string) (GrpcStoreClient, *grpc.ClientConn, error) {
```
Note: I think using the gRPC client pool with the Thanos DNS service discovery (like we did [here](https://github.com/cortexproject/cortex/blob/master/pkg/querier/blocks_store_balanced_set.go)) would have been preferrable here, but **this can be left for a future iteration**, so no need to reiterate on it in this PR.

pracucci commented on pkg/chunk/grpc/grpc_client.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+func connectToGrpcServer(serverAddress string) (GrpcStoreClient, *grpc.ClientConn, error) {
+	cc, err := grpc.Dial(serverAddress, grpc.WithInsecure())
```
Given there's no health checking mechanism (ie. we're not using the client pool) we should setup the gRPC keepalive (we had few outages with etcd due to this). See `grpc.WithKeepaliveParams()` and the [etcd implementation](https://github.com/cortexproject/cortex/blob/c52331d3cbd4d8aa36106f0e2aeb3c1ecf8e0016/vendor/go.etcd.io/etcd/clientv3/client.go#L235-L240) as a reference.

VineethReddy02 replied:
@pracucci Do you have any suggestions on these params. the keepalive check interval and the timeout duration when we don't have active rpcs?
I am thinking to put a 1 min interval to check is alive and 2 mins timeout interval to close the connection if no activity is noticed.
```
	params := keepalive.ClientParameters{
		Time:                time.Minute*1,
		Timeout:           time.Minute*2,
		PermitWithoutStream: true,
	}
```

pracucci replied:
I would be even more aggressive, like: 
- Time: 20s
- Timeout: 10s
- PermitWithoutStream: true

pracucci commented on pkg/chunk/grpc/index_client.go:
```diff
@@ -0,0 +1,98 @@
+}
+
+func (w *WriteBatch) Delete(tableName, hashValue string, rangeValue []byte) {
+	// TODO: implement this to support deleting index entries for gRPC spec
```
We should resolve this TODO.

VineethReddy02 replied:
@pracucci I didn't notice any delete implementation for Cassandra and other stores. As it was TODO in respective store implementations. I reflected the same here. Do you want me to drop such comment with TODO?

pracucci replied:
I think it has been added after you initiated this work. Please rebase `master`.

pracucci commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,124 @@
+	cfg        Config
+	schemaCfg  chunk.SchemaConfig
+	client     GrpcStoreClient
+	connection *grpc.ClientConn
```
Not used. Can be removed.

pracucci replied:
This comment is still valid. `connection` is unused, but I think we should use it to close the connection in `Stop()`.

pracucci commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,124 @@
+func (s *StorageClient) Stop() {
+	_, err := s.client.Stop(context.Background(), &Nothing{})
+	if err != nil {
+		return
```
Never suppress errors. Please log it.

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+
+service grpc_store {
+    //index-client
+    rpc BatchWrite(WriteBatch) returns (Nothing);
```
What's this `Nothing`? Why do we need it?

VineethReddy02 replied:
So in proto files. When we have nothing to return back. We maintain an empty message i.e
`message Nothing {};`
If there any workaround to avoid this. I can consider using them.

pracucci replied:
You can use the built-in `google.protobuf.Empty` ([see this](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/empty.proto))

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+}
+
+message Chunk {
+    bytes buf         = 1;
```
I would name it more clearly. This is the encoded chunk, so why not calling it `encoded`?

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+message Nothing {}
+
+message ChunkID {
+    string chunkId = 1;
```
```suggestion
    string chunkID = 1;
```

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+
+message Nothing {}
+
+message ChunkID {
```
Why do we need a message for this? Looks overkilled to me. A string would have been enough IMO.

VineethReddy02 replied:
So we use ChunkID message only in DeleteChunk rpc as
` rpc DeleteChunks(ChunkID) returns (Nothing); `
I will replace this into  
` rpc DeleteChunks(string) returns (Nothing); `

pracucci replied:
👍

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+    //storage-client
+    rpc PutChunks(Chunks) returns (Nothing);
+    //In GetChunks rpc request send buf as nil
+    rpc GetChunks(Chunks) returns (stream Chunks);
```
I'm not sure `Chunks` is the right data structure to use here. Design wise, it would be cleaner if we would have a different data structure for read and write. In the read path (this rpc) what you actually need is a list of chunk keys (table name and key). In the write path you need the chunk key + data.

VineethReddy02 replied:
Initially, we maintained different structures but @gouthamve, and I felt to use the same data structure for read & write path to make it simple. Maybe @gouthamve can here on the context behind this decision.

pracucci replied:
Design wise I think it's a bit risky for future evolutions. As I commented in the review message, I think that having a dedicated Request and Response message for each function is more future proof, ie.

```
rpc PutChunks(PutChunksRequest) returns (google.protobuf.Empty);
rpc GetChunks(GetChunksRequest) returns (stream GetChunksResponse);
```

pracucci commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,124 @@
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		if receivedChunks != nil {
```
You can remove this if and use `range receivedChunks.GetChunks()` in the next line.

pracucci commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,124 @@
+		if receivedChunks != nil {
+			for _, chunkResponse := range receivedChunks.Chunks {
+				var c chunk.Chunk
+				if chunkResponse != nil {
```
When this could happen?

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+	connection *grpc.ClientConn
+}
+
+// NewStorageClient returns a new StorageClient.
```
```suggestion
// NewTableClient returns a new StorageClient.
```

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+type TableClient struct {
+	cfg        Config
+	client     GrpcStoreClient
+	connection *grpc.ClientConn
```
Unused. Please remove it.

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// NewStorageClient returns a new StorageClient.
+func NewTableClient(ctx context.Context, cfg Config) (*TableClient, error) {
```
`ctx` is unused. Please remove it.

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+	stdContext "context"
+
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
```
I don't think this should be used. Please use only `"context"`.

pracucci commented on pkg/chunk/grpc/gRPC.proto:
```diff
@@ -0,0 +1,111 @@
+}
+
+message ListTablesResponse {
+    repeated string tables = 1;
```
```suggestion
    repeated string tableNames = 1;
```

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+	desc.ProvisionedWrite = tableDesc.Desc.ProvisionedWrite
+	desc.UseOnDemandIOMode = tableDesc.Desc.UseOnDemandIOMode
+	desc.Tags = tableDesc.Desc.Tags
+	return desc, true, nil
```
Why we always return `true` as `isActive`? The response contains `IsActive` which we ignore.

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+	tableName := &TableName{TableName: name}
+	tableDesc, err := c.client.DescribeTable(ctx, tableName)
+	if err != nil {
+		return desc, true, errors.WithStack(err)
```
A minor thing, but a good practice when returning errors:

```suggestion
		return desc, false, errors.WithStack(err)
```

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,113 @@
+	currentTable.UseOnDemandIOMode = current.UseOnDemandIOMode
+	currentTable.ProvisionedWrite = current.ProvisionedWrite
+	currentTable.ProvisionedRead = current.ProvisionedRead
+	if current.Tags != nil {
+		currentTable.Tags = current.Tags
+	} else {
+		currentTable.Tags = map[string]string{}
+	}
```
Why don't just allow to set `nil` to `currentTable`? To give you an example, in `CreateTable()` you're not doing this and you allow to set `Tags` to `nil`.

VineethReddy02 commented:
@pracucci @gouthamve 
Addressed the review comments & added unit tests by writing the gRPC mock server.
Feel free to suggest changes around unit tests. I am writing unit tests with a mock gRPC server for the first time. :) Would love to iterate it if needed.
I will also extend that with integration tests for this feature in the future PR's.

gouthamve commented on pkg/chunk/grpc/grpc_client_test.go:
```diff
@@ -0,0 +1,183 @@
+
+	"context"
+
+	"github.com/prometheus/prometheus/pkg/labels"
```
Why are we spacing the imports out?

gouthamve commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,207 @@
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
```
Same here

VineethReddy02 replied:
In all the go files in the cortex code base, I see this grouping of imports I mean standards packages, other Github imports, and then the cortex project reference imports.

gouthamve commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -0,0 +1,100 @@
+
+import (
+	"context"
+
```
Same here

gouthamve commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,128 @@
+					return result, err
+				}
+			}
+			if c.Data != nil {
```
I think we can remove this check here.

VineethReddy02 replied:
@gouthamve for some reason if the gRPC server sends `c.Data `stream as nil cortex will crash.
So it's an additional check to ensure we append data which is not nil. If it's okay I can remove it.

pracucci replied:
I would remove it as well. Silently ignoring this condition may lead to very subdle bugs. Having Cortex panicking sounds actually better in this edge case.

gouthamve commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,128 @@
+
+	"context"
+
+	"github.com/pkg/errors"
```
Don't space imports here too

pracucci commented on pkg/chunk/grpc/storage_client.go:
```diff
@@ -0,0 +1,124 @@
+		for _, chunkResponse := range receivedChunks.GetChunks() {
+			var c chunk.Chunk
+			if chunkResponse != nil {
+				decodeContext = chunk.NewDecodeContext()
```
We've recently learned this is an expensive operation. You can use the same `decodeContext` to decode all chunks, because `c.Decode()` will call `Reset()` on the decode context before decoding the chunk.

pracucci commented on pkg/chunk/grpc/grpc.proto:
```diff
@@ -0,0 +1,121 @@
+    string chunkID = 1;
+}
+
+message TableName {
```
A message `TableName` containing `tableName` looks a weird design to me. If you wanna keep it extensible, then I would split it into:

```
message DeleteTableRequest {
  string tableName = 1;
}

message DescribeTableRequest {
  string tableName = 1;
}
```

pracucci commented on pkg/chunk/grpc/grpc.proto:
```diff
@@ -0,0 +1,121 @@
+service grpc_store {
+    //index-client
+    rpc BatchWrite(BatchWrites) returns (google.protobuf.Empty);
+    rpc QueryPages(IndexQuery) returns (stream ReadBatch);
```
Should we call it `QueryIndex(QueryIndexRequest) returns (stream QueryIndexResponse)` for clarity/consistency?

pracucci commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,203 @@
+	return NewGrpcStoreClient(cc), cc, nil
+}
+
+func createTestGrpcServer() {
```
You can pass `t *testing.T` and then assert on no error using `require.NoError(t, err)`. That's a pattern we frequently use. For example:

```
lis, err := net.Listen("tcp", "localhost:6688")
require.NoError(t, err)
```

pracucci commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,203 @@
+	return client, nil
+}
+
+func connectToTestGrpcServer(serverAddress string) (GrpcStoreClient, *grpc.ClientConn, error) {
```
How is this different than `connectToGrpcServer()`? Why not using `connectToGrpcServer()`?

VineethReddy02 replied:
yeah. We can use it.

pracucci commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,203 @@
+
+func createTestGrpcServer() {
+	var cfg server
+	lis, err := net.Listen("tcp", "localhost:6688")
```
Don't hardcode ports. It makes tests more flaky. You can listen on `localhost:0`, and then having `createTestGrpcServer()` returning the address where the server is listening to.

pracucci commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,203 @@
+		log.Fatalf("failed to listen: %v", err)
+	}
+	s := grpc.NewServer()
+	// Enable reflection to expose endpoints this server offers.
+	reflection.Register(s)
```
Why do you need this?

VineethReddy02 replied:
Yeah, we don't need this. Didn't notice.

pracucci commented on pkg/chunk/grpc/grpc_server_mock.go:
```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+	s := grpc.NewServer()
```
We should call `GracefulStop()` at the end of the test. A tecnique you can use is returning a cleanup function from `createTestGrpcServer()` and then having the test calling the cleanup in a defer. For example:

```
func createTestGrpcServer() (func()) {
  s := grpc.NewServer()
  cleanup := func() {
    s.GracefulStop()
  }

  return cleanup
```

_This is pattern we frequently use._

VineethReddy02 commented on pkg/chunk/grpc/grpc.proto:
```diff
@@ -0,0 +1,121 @@
+    rpc UpdateTable(UpdateTableRequest) returns (google.protobuf.Empty);
+
+    //used by storage-client & index-client
+    rpc Stop(google.protobuf.Empty) returns (google.protobuf.Empty);
```
So, Initially I was thinking there is no use of Stop(). But I was thinking in next iteration when we add support for multiple grpc-stores. If for some reason we want to terminate the connection from the client-side with one of the grpc-store maybe this can help, just a thought. We can definitely drop this. If we have no use.

pracucci replied:
I would suggest to remove `Stop()` and add it if/when required.

VineethReddy02 commented:
Thanks @pracucci for taking the time & reviewing. I made all the suggested changes. 
Looking forward in getting this merged. :)

pracucci commented on pkg/chunk/grpc/grpc.proto:
```diff
@@ -0,0 +1,122 @@
+
+    //table-client
+    rpc ListTables(google.protobuf.Empty) returns (ListTablesResponse);
+    rpc CreateTable(TableDesc) returns (google.protobuf.Empty);
```
I would define `CreateTableRequest` for extensibility, like we did for any other rpc. The `CreateTableRequest` can just contain `TableDesc` for now.

VineethReddy02 replied:
Rather nesting the rpc messages. I think it's better to rename it as suggested.

pracucci replied:
Yes, but `TableDesc` is also used in other places, no?

VineethReddy02 replied:
Ahh... I didn't notice. TableDesc being used in many other places. I will nest the message as suggested.

pracucci commented on pkg/chunk/grpc/table_client.go:
```diff
@@ -84,12 +84,13 @@ func (c *TableClient) UpdateTable(ctx context.Context, current, expected chunk.T
-	tableDesc.ProvisionedWrite = desc.ProvisionedWrite
-	tableDesc.Tags = desc.Tags
-	tableDesc.UseOnDemandIOMode = desc.UseOnDemandIOMode
+	tableDesc := &CreateTableRequest{}
```
Please rename `tableDesc` to `req`, otherwise below it's a bit weird reading `tableDesc.Desc.*`.

pstibrany commented:
Since this API is supposedly to be used for writing 3rd party storage implementations, I think it lacks documentation (at least inside grpc.proto file, but more generic documentation about how Cortex uses the storage and what should the user of this API reasonably expect).

Is there any implementation using this API? How do we know it satisfies the needs of external storage client?

(I am also skeptical about including this in Cortex in the first place. Many failure modes related to gRPC (grpc server not running, big responses, bad config, ...) would be automatically removed if client was included in the Cortex)

gouthamve commented:
> I think it lacks documentation

Absolutely. We still have follow up work to document and give a reference in-memory implementation. And to use the in-mem implementation to write an integration test.

> would be automatically removed if client was included in the Cortex

What do you mean?

pstibrany commented:
> Absolutely. We still have follow up work to document and give a reference in-memory implementation. And to use the in-mem implementation to write an integration test.

I think we need real implementation or two, not just in-mem one.

> > would be automatically removed if client was included in the Cortex
> 
> What do you mean?

I was talking about failure modes related to gRPC -- which wouldn't be a problem if client was in Cortex directly, and not using gRPC (well, obviously). Just saying that using gRPC adds more complexity.

gouthamve commented:
I kinda disagree :) I see gRPC having the same amount of failure modes as a remote Bigtable/Cassandra. They are both endpoints.

> grpc server not running, big responses, bad config, ...

We faced these with Bigtable tbh, specially the big responses. We had to up our gRPC message limit with a previous schema, hence came up with the v9/v10 schema. I would say gRPC makes the failure modes more likely, but doesn't introduce new ones.

And while I agree that having the client directly in Cortex would be the best, I am not super enthused about a ton of storage clients in the tree.

gouthamve commented:
> I think we need real implementation or two, not just in-mem one.

The first one was going to be Postgres/MySQL so Loki could use it as the index. Not sure what the next one would be tbh.

pstibrany commented:
> And while I agree that having the client directly in Cortex would be the best, I am not super enthused about a ton of storage clients in the tree.

I personally think that biggest issue in people contributing more clients is lack of good documentation, and pointers about how to start.

Anyway, please feel free to merge when PR is ready, my comments are not merge-blockers.

VineethReddy02 commented:
@gouthamve @pracucci 
I see CI fails with quay error.

`Error response from daemon: Get https://quay.io/v2/: socks connect tcp 127.0.0.1:1080->quay.io:443: unknown error address type not supported`

pracucci commented:
Let's gooo 🚀 ! _I will merge as soon as the CI passes_
